File: AutoRest/Generators/NodeJS/NodeJS.Tests/AcceptanceTests/headerTests.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

'use strict';

import should = require('should');
import http = require('http');
import util = require('util');
import assert = require('assert');
import msRest = require('ms-rest');
import moment = require('moment');
var _ = require('underscore');

import headerClient = require('../Expected/AcceptanceTests/Header/autoRestSwaggerBATHeaderService');

var dummyToken = 'dummy12321343423';
var credentials = new msRest.TokenCredentials(dummyToken);

var clientOptions = {};
var baseUri = 'http://localhost:3000';

describe('nodejs', function () {

  describe('Swagger Header BAT', function () {

    describe('Basic Header Operations', function () {
      var testClient = new headerClient(baseUri, clientOptions);
      it('should override existing headers', function (done) {
        testClient.header.paramExistingKey('overwrite', function (error, result, request, response) {
          should.not.exist(error);
          testClient.header.responseExistingKey(function (error, result, request, response) {
            response.headers['user-agent'].should.be.exactly('overwrite');
            done();
          });
        });
      });

      it('should throw on changing protected headers', function(done) {
        testClient.header.paramProtectedKey('text/html', function (error, result, request, response) {
          should.not.exist(error);
          testClient.header.responseProtectedKey(function (error, result, request, response) {
            response.headers['content-type'].should.be.exactly('text/html; charset=utf-8');
            done();
          });
        });
      });

      it('should send and receive integer type headers', function(done) {
        testClient.header.paramInteger('positive', 1, function(error, result) {
          should.not.exist(error);
          testClient.header.paramInteger('negative', -2, function (error, result) {
            should.not.exist(error);
            testClient.header.responseInteger('positive', function (error, result, request, response) {
              should.not.exist(error);
              response.headers['value'].should.be.exactly('1');
              testClient.header.responseInteger('negative', function (error, result, request, response) {
                should.not.exist(error);
                response.headers['value'].should.be.exactly('-2');
                done();
              });
            });
          });
        });
      });

      it('should send and receive long type headers', function(done) {
        testClient.header.paramLong('positive', 105, function(error, result) {
          should.not.exist(error);
          testClient.header.paramLong('negative', -2, function (error, result) {
            should.not.exist(error);
            testClient.header.responseLong('positive', function (error, result, request, response) {
              should.not.exist(error);
              response.headers['value'].should.be.exactly('105');
              testClient.header.responseLong('negative', function (error, result, request, response) {
                should.not.exist(error);
                response.headers['value'].should.be.exactly('-2');
                done();
              });
            });
          });
        });
      });

      it('should send and receive float type headers', function(done) {
        testClient.header.paramFloat('positive', 0.07, function(error, result) {
          should.not.exist(error);
          testClient.header.paramFloat('negative', -3.0, function (error, result) {
            should.not.exist(error);
            testClient.header.responseFloat('positive', function (error, result, request, response) {
              should.not.exist(error);
              response.headers['value'].should.be.exactly('0.07');
              testClient.header.responseFloat('negative', function (error, result, request, response) {
                should.not.exist(error);
                JSON.parse(response.headers['value']).should.be.exactly(-3.0);
                done();
              });
            });
          });
        });
      });

      it('should send and receive double type headers', function(done) {
        testClient.header.paramDouble('positive', 7e120, function(error, result) {
          should.not.exist(error);
          testClient.header.paramDouble('negative', -3.0, function (error, result) {
            should.not.exist(error);
            testClient.header.responseDouble('positive', function (error, result, request, response) {
              should.not.exist(error);
              JSON.parse(response.headers['value']).should.be.exactly(7e120);
              testClient.header.responseDouble('negative', function (error, result, request, response) {
                should.not.exist(error);
                JSON.parse(response.headers['value']).should.be.exactly(-3.0);
                done();
              });
            });
          });
        });
      });

      it('should send and receive boolean type headers', function(done) {
        testClient.header.paramBool('true', true, function(error, result) {
          should.not.exist(error);
          testClient.header.paramBool('false', false, function (error, result) {
            should.not.exist(error);
            testClient.header.responseBool('true', function (error, result, request, response) {
              should.not.exist(error);
              response.headers['value'].should.be.exactly('true');
              testClient.header.responseBool('false', function (error, result, request, response) {
                should.not.exist(error);
                response.headers['value'].should.be.exactly('false');
                done();
              });
            });
          });
        });
      });
      it('should send and receive string type headers', function (done) {
        testClient.header.paramString('valid', <any>{ value: 'The quick brown fox jumps over the lazy dog' }, function (error, result) {
          should.not.exist(error);
          testClient.header.paramString('null', { value: null }, function (error, result) {
            should.not.exist(error);
            testClient.header.paramString('empty', { value: '' }, function (error, result) {
              should.not.exist(error);
              testClient.header.responseString('valid', function (error, result, request, response) {
                should.not.exist(error);
                response.headers['value'].should.be.exactly('The quick brown fox jumps over the lazy dog');
                testClient.header.responseString('null', function (error, result, request, response) {
                  should.not.exist(error);
                  should.not.exist(JSON.parse(response.headers['value']));
                  testClient.header.responseString('empty', function (error, result, request, response) {
                    should.not.exist(error);
                    response.headers['value'].should.be.exactly('');
                    done();
                  });
                });
              });
            });
          });
        });
      });
      it('should send and receive enum type headers', function (done) {
        testClient.header.paramEnum('valid', { value: 'GREY' }, function (error, result) {
          should.not.exist(error);
          testClient.header.paramEnum('null', { value: null }, function (error, result) {
            should.not.exist(error);
            testClient.header.responseEnum('valid', function (error, result, request, response) {
              should.not.exist(error);
              response.headers['value'].should.be.exactly('GREY');
              testClient.header.responseEnum('null', function (error, result, request, response) {
                should.not.exist(error);
                response.headers['value'].should.be.exactly('');
                done();
              });
            });
          });
        });
      });
      it('should send and receive date type headers', function (done) {
        testClient.header.paramDate('valid', new Date('2010-01-01'), function (error, result) {
          should.not.exist(error);
          testClient.header.paramDate('min', new Date('0001-01-01'), function (error, result) {
            should.not.exist(error);
            testClient.header.responseDate('valid', function (error, result, request, response) {
              should.not.exist(error);
              _.isEqual(new Date(response.headers['value']), new Date('2010-01-01')).should.be.exactly(true);
              testClient.header.responseDate('min', function (error, result, request, response) {
                should.not.exist(error);
                _.isEqual(new Date(response.headers['value']), new Date('0001-01-01')).should.be.exactly(true);
                done();
              });
            });
          });
        });
      });
      it('should send and receive datetime type headers', function (done) {
        testClient.header.paramDatetime('valid', new Date('2010-01-01T12:34:56Z'), function (error, result) {
          should.not.exist(error);
          testClient.header.paramDatetime('min', new Date('0001-01-01T00:00:00Z'), function (error, result) {
            should.not.exist(error);
            testClient.header.responseDatetime('valid', function (error, result, request, response) {
              should.not.exist(error);
              _.isEqual(new Date(response.headers['value']), new Date('2010-01-01T12:34:56Z')).should.be.exactly(true);
              testClient.header.responseDatetime('min', function (error, result, request, response) {
                should.not.exist(error);
                _.isEqual(new Date(response.headers['value']), new Date('0001-01-01T00:00:00Z')).should.be.exactly(true);
                done();
              });
            });
          });
        });
      });
      it('should send and receive datetimerfc1123 type headers', function (done) {
        testClient.header.paramDatetimeRfc1123('valid', <any>{ value: new Date('2010-01-01T12:34:56Z') }, function (error, result) {
          should.not.exist(error);
          testClient.header.paramDatetimeRfc1123('min', { value: new Date('0001-01-01T00:00:00Z') }, function (error, result) {
            should.not.exist(error);
            testClient.header.responseDatetimeRfc1123('valid', function (error, result, request, response) {
              should.not.exist(error);
              _.isEqual(new Date(response.headers['value']), new Date('Fri, 01 Jan 2010 12:34:56 GMT')).should.be.exactly(true);
              testClient.header.responseDatetimeRfc1123('min', function (error, result, request, response) {
                should.not.exist(error);
                _.isEqual(new Date(response.headers['value']), new Date('Mon, 01 Jan 0001 00:00:00 GMT')).should.be.exactly(true);
                done();
              });
            });
          });
        });
      });
      it('should send and receive duration type headers', function (done) {
        var duration = moment.duration({ days: 123, hours: 22, minutes: 14, seconds: 12, milliseconds: 11 });
        testClient.header.paramDuration('valid', duration, function (error, result) {
          should.not.exist(error);
          testClient.header.responseDuration('valid', function (error, result, request, response) {
            should.not.exist(error);
            _.isEqual(response.headers['value'], 'P123DT22H14M12.011S').should.be.exactly(true);
            done();
          });
        });
      });
      it('should send and receive byte array type headers', function (done) {
        var bytes = new Buffer('啊齄丂狛狜隣郎隣兀﨩');
        testClient.header.paramByte('valid', bytes, function (error, result) {
          should.not.exist(error);
          testClient.header.responseByte('valid', function (error, result, request, response) {
            should.not.exist(error);
            response.headers['value'].should.be.exactly(bytes.toString('base64'));
            done();
          });
        });
      });
    });
  });
});

File: src/generator/AutoRest.Python.Tests/AcceptanceTests/header_tests.py
# --------------------------------------------------------------------------
#
# Copyright (c) Microsoft Corporation. All rights reserved.
#
# The MIT License (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the ""Software""), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
#
# --------------------------------------------------------------------------

import unittest
import subprocess
import sys
import isodate
import tempfile
import json
from datetime import date, datetime, timedelta
import os
from os.path import dirname, pardir, join, realpath

cwd = dirname(realpath(__file__))
log_level = int(os.environ.get('PythonLogLevel', 30))

tests = realpath(join(cwd, pardir, "Expected", "AcceptanceTests"))
sys.path.append(join(tests, "Header"))

from msrest.serialization import Deserializer
from msrest.exceptions import DeserializationError

from auto_rest_swagger_bat_header_service import AutoRestSwaggerBATHeaderService
from auto_rest_swagger_bat_header_service.models.auto_rest_swagger_bat_header_service_enums import GreyscaleColors


class HeaderTests(unittest.TestCase):

    def test_headers(self):
        client = AutoRestSwaggerBATHeaderService(base_url="http://localhost:3000")

        client.header.param_integer("positive", 1)
        client.header.param_integer("negative", -2)

        raw = client.header.response_integer("positive", raw=True)
        self.assertEqual(1, int(raw.response.headers.get("value")))
        self.assertEqual(1, raw.headers.get("value"))
        raw = client.header.response_integer("negative", raw=True)
        self.assertEqual(-2, raw.headers.get("value"))

        client.header.param_long("positive", 105)
        client.header.param_long("negative", -2)

        raw = client.header.response_long("positive", raw=True)
        self.assertEqual(105, raw.headers.get("value"))
        raw = client.header.response_long("negative", raw=True)
        self.assertEqual(-2, raw.headers.get("value"))

        client.header.param_float("positive", 0.07)
        client.header.param_float("negative", -3.0)

        raw = client.header.response_float("positive", raw=True)
        self.assertTrue(abs(0.07 - raw.headers.get("value")) < 0.00001)
        raw = client.header.response_float("negative", raw=True)
        self.assertTrue(abs(-3.0 - raw.headers.get("value")) < 0.00001)

        client.header.param_double("positive", 7e120)
        client.header.param_double("negative", -3.0)

        raw = client.header.response_double("positive", raw=True)
        self.assertEqual(7e120, raw.headers.get("value"))
        raw = client.header.response_double("negative", raw=True)
        self.assertEqual(-3.0, raw.headers.get("value"))

        client.header.param_bool("true", True)
        client.header.param_bool("false", False)

        raw = client.header.response_bool("true", raw=True)
        self.assertEqual(True, raw.headers.get("value"))
        raw = client.header.response_bool("false", raw=True)
        self.assertEqual(False, raw.headers.get("value"))

        client.header.param_string("valid", "The quick brown fox jumps over the lazy dog")
        client.header.param_string("null", None)
        client.header.param_string("empty", "")

        raw = client.header.response_string("valid", raw=True)
        self.assertEqual("The quick brown fox jumps over the lazy dog", raw.headers.get("value"))
        raw = client.header.response_string("null", raw=True)
        self.assertEqual(None, json.loads(raw.headers.get("value")))
        raw = client.header.response_string("empty", raw=True)
        self.assertEqual("", raw.headers.get("value"))

        client.header.param_enum("valid", GreyscaleColors.grey)
        client.header.param_enum("valid", 'GREY')
        client.header.param_enum("null", None)

        raw = client.header.response_enum("valid", raw=True)
        self.assertEqual(GreyscaleColors.grey, raw.headers.get("value"))

        # We can't deserialize 'null'
        with self.assertRaises(DeserializationError):
            raw = client.header.response_enum("null", raw=True)

        client.header.param_date("valid", isodate.parse_date("2010-01-01"))
        client.header.param_date("min", datetime.min)

        raw = client.header.response_date("valid", raw=True)
        self.assertEqual(isodate.parse_date("2010-01-01"), raw.headers.get("value"))
        raw = client.header.response_date("min", raw=True)
        self.assertEqual(isodate.parse_date("0001-01-01"), raw.headers.get("value"))

        client.header.param_datetime("valid", isodate.parse_datetime("2010-01-01T12:34:56Z"))
        client.header.param_datetime("min", datetime.min)

        raw = client.header.response_datetime("valid", raw=True)
        self.assertEqual(isodate.parse_datetime("2010-01-01T12:34:56Z"), raw.headers.get("value"))
        raw = client.header.response_datetime("min", raw=True)
        self.assertEqual(isodate.parse_datetime("0001-01-01T00:00:00Z"), raw.headers.get("value"))

        client.header.param_datetime_rfc1123("valid", isodate.parse_datetime("2010-01-01T12:34:56Z"))
        client.header.param_datetime_rfc1123("min", datetime.min)

        raw = client.header.response_datetime_rfc1123("valid", raw=True)
        self.assertEqual(isodate.parse_datetime("2010-01-01T12:34:56Z"), raw.headers.get("value"))
        raw = client.header.response_datetime_rfc1123("min", raw=True)
        self.assertEqual(isodate.parse_datetime("0001-01-01T00:00:00Z"), raw.headers.get("value"))

        client.header.param_duration("valid", timedelta(days=123, hours=22, minutes=14, seconds=12, milliseconds=11))

        raw = client.header.response_duration("valid", raw=True)
        self.assertEqual(timedelta(days=123, hours=22, minutes=14, seconds=12, milliseconds=11), raw.headers.get("value"))

        u_bytes = bytearray(u"\u554A\u9F44\u4E02\u72DB\u72DC\uF9F1\uF92C\uF9F1\uFA0C\uFA29", encoding='utf-8')
        client.header.param_byte("valid", u_bytes)

        raw = client.header.response_byte("valid", raw=True)
        self.assertEqual(u_bytes, raw.headers.get("value"))

        client.header.param_existing_key("overwrite")

        raw = client.header.response_existing_key(raw=True)
        self.assertEqual("overwrite", raw.headers.get('User-Agent'))

        # This test is only valid for C#, which content-type can't be override this way
        #client.header.param_protected_key("text/html")

        # This test has different result compare to C#, which content-type is saved in another place.
        raw = client.header.response_protected_key(raw=True)
        self.assertTrue("text/html; charset=utf-8", raw.headers.get('Content-Type'))

        custom_headers = {"x-ms-client-request-id": "9C4D50EE-2D56-4CD3-8152-34347DC9F2B0"}
        raw = client.header.custom_request_id(custom_headers, raw=True)
        self.assertEqual(raw.response.status_code, 200)


if __name__ == '__main__':
    unittest.main()
